### PR TITLE
proxy: Move Kafka proxy to Envoy Go Extensions

### PR DESCRIPTION
**Pull request by jrajahalme**
Move Kafka proxy to Envoy Go Extensions. This has couple of benefits:

- Conceptually simplify Cilium Agent and make it a little bit smaller
  - Removal of some vendored Kafka proxy code now replaced by Envoy/proxylib
- Egress policy support for Kafka proxying
- Performance should be better
  - Envoy threading model is better suited for throughput and low latency, no go-routines are needed
  - Only 8 bytes are parsed from responses, rest is passed on without even passing the data to the parser

The only user-visible change to existing behavior should be the small change in access logging format, and the fact that only CorrelationID is logged for responses.

As an internal implementation detail the generic L7 key-value pair policy format is used to carry Kafka policies from Cilium agent to Envoy. Correspondingly the generic L7 access log format is used to emit the access log messages from the Kafka parser.

```release-note
Kafka proxy is now implemented in Cilium Go extensions for Envoy,
which adds egress policy enforcement support for Kafka L7 policies.
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/hubble/parser/seven/parser.go [hunk 1]
```diff
@@ -395,11 +396,16 @@ func decodeLayer7(r *accesslog.LogRecord) *pb.Layer7 {
...
 		return &pb.Layer7{
 			Type:   flowType,
-			Record: decodeKafka(r.Type, r.Kafka),
+			Record: decodeKafka(r.Type, r.L7),
 		}
+	// case r.L7 != nil:
+	// 	return &pb.Layer7{
+	// 		Type:   flowType,
+	// 		Record: decodeL7(r.Type, r.L7),
+	//	}
```
Comment (gandro) [inline pkg/hubble/parser/seven/parser.go]: This looks like a leftover?

File: pkg/monitor/logrecord.go [hunk 2]
```diff
@@ -167,7 +159,6 @@ type LogRecordNotifyVerbose struct {
 	DstIdentity      uint64                     `json:"dstIdentity"`
 	Verdict          accesslog.FlowVerdict      `json:"verdict"`
 	HTTP             *accesslog.LogRecordHTTP   `json:"http,omitempty"`
-	Kafka            *accesslog.LogRecordKafka  `json:"kafka,omitempty"`
```
Comment (gandro) [inline pkg/monitor/logrecord.go]: This is technically a breaking change in the monitor API. Not opposed to this, but just wanted to point that out explicitly, e.g. legacy versions of Hubble (v0.5) would not be able to recognize Kafka records after this (we don't support Hubble v0.5 on Cilium 1.9, so that's fine).

**Review by gandro — APPROVED** [review]
Hubble-related changes look good to me! One comment with regards to the monitor API.

I did not review the policy or parser changes.

File: pkg/hubble/parser/seven/parser.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/hubble/parser/seven/parser.go]: More like wishful thinking... it seems to me that the generic L7 key-value pairs are not translated to the monitor API yet?

File: pkg/monitor/logrecord.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/monitor/logrecord.go]: Good to know, will leave this for further discussion.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Added two commits to exercise Kafka egress policies.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-gke

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/hubble/parser/seven/parser.go [hunk 3]
```diff
@@ -513,24 +519,28 @@ func (p *Parser) getSummary(logRecord *accesslog.LogRecord, flow *pb.Flow) strin
...
-			ErrorCode:     int32(kafka.ErrorCode),
-			ApiVersion:    int32(kafka.APIVersion),
-			ApiKey:        kafka.APIKey,
-			CorrelationId: kafka.CorrelationID,
-			Topic:         kafka.Topic.Topic,
+			ErrorCode:     int32(errorCode),
+			ApiVersion:    int32(apiVersion),
+			ApiKey:        l7.Fields["APIKey"],
+			CorrelationId: int32(correlationID),
+			Topic:         l7.Fields["Topics"],
```
Comment (aanm) [inline pkg/hubble/parser/seven/parser.go]: Should we have constants for these keys? Also, shouldn't this `decodeKafka` function be in a dedicated file, perhaps `pkg/hubble/parser/seven/kafka.go`?

File: proxylib/kafka/parser.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+	if l7Rules == nil {
+		return nil
+	}
+
+	allowRules := l7Rules.GetL7Rules()
+	rules := make([]L7NetworkPolicyRule, 0, len(allowRules))
+	for _, l7Rule := range allowRules {
+		var kr KafkaRule
+		for k, v := range l7Rule.Rule {
+			switch k {
```
Comment (aanm) [inline proxylib/kafka/parser.go]: Should we have `switch strings.ToLower(k)` {?

File: vendor/modules.txt [hunk 5]
```diff
@@ -475,7 +475,6 @@ github.com/opencontainers/image-spec/specs-go
 github.com/opencontainers/image-spec/specs-go/v1
```
Comment (aanm) [inline vendor/modules.txt:481]: Given that we are only using the `kafka/proto` we can remove the replace directive for this dependency in `go.mod` and use the one from upstream instead our fork?

File: proxylib/kafka/parser.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+		req.topics[topic] = struct{}{}
+	}
+	logEntry := &cilium.LogEntry_GenericL7{
+		GenericL7: &cilium.L7LogEntry{
+			Proto: parserName,
+			Fields: map[string]string{
+				"APIVersion":    strconv.Itoa(int(req.GetVersion())),
+				"APIKey":        apiKeyToString(req.GetAPIKey()),
+				"CorrelationID": strconv.Itoa(int(req.GetCorrelationID())),
+				"Topics":        strings.Join(topics, ","),
```
Comment (aanm) [inline proxylib/kafka/parser.go]: these could be constants as well

**Review by aanm — CHANGES_REQUESTED** [review]

File: proxylib/kafka/parser.go (hunk 4 above)
Comment (jrajahalme) [reply proxylib/kafka/parser.go]: Not necessarily as these keys are generated by own code in Cilium agent which currently uses all lowercase letters.

File: vendor/modules.txt (hunk 5 above)
Comment (jrajahalme) [reply vendor/modules.txt:481]: I was not aware of the replace directive, will check!

Comment (jrajahalme) [reply vendor/modules.txt:481]: @aanm The three commits we have on our fork a head of the upstream ancestor are all on `kafka/proto`. Now that we are not parsing requests at all some of there commits may not be needed any more, but the latest commit fixes an issue in request parsing that seems like we'd want to keep.

File: pkg/hubble/parser/seven/parser.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/hubble/parser/seven/parser.go]: Dedicated file or not is not an issue for this PR as I did not introduce the function here.

File: proxylib/kafka/parser.go (hunk 6 above)
Comment (jrajahalme) [reply proxylib/kafka/parser.go]: Will await for some more feedback on the general approach of using the "generic L7" format to internally carry Kafka log messages before doubling down on it :-)

**Comment by jrajahalme** [comment]
Coverage has decreased due to the removed `proxy/kafka` code having been some of the better covered code...

File: pkg/kafka/policy.go [hunk 7]
```diff
@@ -141,15 +141,14 @@ func matchOffsetFetchReq(req *proto.OffsetFetchReq, rule api.PortRuleKafka) bool
...
 	if req == nil {
 		return false
 	}
 
-	flowdebug.Log(log.WithFields(logrus.Fields{
-		fieldRequest: req.String(),
-		fieldRule:    rule,
-	}), "Matching Kafka rule")
+	if flowdebug.Enabled() {
+		log.Debugf("Matching Kafka request %s against rule %v", req.String(), *rule)
```
Comment (nebril) [inline pkg/kafka/policy.go]: Why the change to logging method? Seems like structured log was nicer, would it be possible to get back to it?

File: pkg/kafka/request.go [hunk 8]
```diff
@@ -217,12 +223,15 @@ func ReadRequest(reader io.Reader) (*RequestMessage, error) {
 	case proto.OffsetFetchReqKind:
 		req.request, err = proto.ReadOffsetFetchReq(buf)
 	default:
-		log.WithField(fieldRequest, req.String()).Debugf("Unknown Kafka request API key: %d", req.kind)
+		if flowdebug.Enabled() {
+			log.Debugf("Unknown Kafka request API key: %d in %s", req.kind, req.String())
```
Comment (nebril) [inline pkg/kafka/request.go:255]: Same as comment above, please consider using structured log, (but `if flowdebug.Enabled` line is probably a good call).

Will not comment on more changes to structured logs to not clutter the review.

File: pkg/monitor/logrecord.go (hunk 2 above)
Comment (nebril) [reply pkg/monitor/logrecord.go]: I think it's fine to go through with that, but would be nice to have a release note. If it's necessary to leave this, can we have information about Kafka in two places (AFAIU right now it will be stored in `L7` field?).

**Review by nebril — APPROVED** [review]
Aside from logging nitpicks I don't have any issues. Great work!

**Review by aanm — APPROVED** [review]
Nothing else besides what Maciej already point it out.

File: pkg/kafka/policy.go (hunk 7 above)
Comment (jrajahalme) [reply pkg/kafka/policy.go]: Short answer: to avoid confusing log messages.

Long answer: Kafka logs become part of Envoy logs when Kafka is moved to the proxylib. Envoy log messages are not structured and currently we have no code to try parse the structured items from the main log message. This means that if we use structured logs here, they will be embedded in to the main log message and will not be found by any code that will try to parse structured logs. For reference please see https://github.com/cilium/cilium/blob/master/pkg/envoy/envoy.go#L282-L343

File: pkg/kafka/request.go (hunk 8 above)
Comment (jrajahalme) [reply pkg/kafka/request.go:255]: See above :-)

File: pkg/monitor/logrecord.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/monitor/logrecord.go]: I'll check how hard it would be to keep using `LogRecordKafka`, will report back!

File: pkg/hubble/parser/seven/parser.go [hunk 9]
```diff
@@ -395,11 +396,16 @@ func decodeLayer7(r *accesslog.LogRecord) *pb.Layer7 {
 			Type:   flowType,
 			Record: decodeHTTP(r.Type, r.HTTP),
 		}
-	case r.Kafka != nil:
+	case r.L7 != nil && r.L7.Proto == "kafka":
```
Comment (aditighag) [inline pkg/hubble/parser/seven/parser.go]: minor: Should we extract "kafka" from here and L 511 to a constant?

**Review by aditighag — APPROVED** [review]

File: proxylib/kafka/parser.go [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+	if length < framelength {
+		// Not enough data, ask for more and try again
+		return MORE, framelength - length
+	}
+
+	var req KafkaRequest
+	req.RequestMessage, err = kafka.ReadRequest(reader)
+	if err != nil {
+		if flowdebug.Enabled() {
+			log.WithError(err).Warning("Unable to parse Kafka request; closing Kafka connection")
```
Comment (aditighag) [inline proxylib/kafka/parser.go:105]: Does it print metadata (e.g., correlation ID) for the request in the log output?

File: proxylib/kafka/parser.go [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+	if p.connection.Matches(req) {
+		p.connection.Log(cilium.EntryType_Request, logEntry)
+		return PASS, framelength
+	}
+
+	resp, err := req.CreateAuthErrorResponse()
+	if err != nil {
+		logEntry.GenericL7.Fields["status"] = strconv.Itoa(int(kafka.ErrInvalidMessage))
+		if flowdebug.Enabled() {
+			log.WithError(err).Warning("Unable to create Kafka response")
```
Comment (aditighag) [inline proxylib/kafka/parser.go:130]: Same comment as L 163.

**Review by aditighag — APPROVED** [review]

File: pkg/policy/api/kafka/doc.go [hunk 12]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017 Authors of Cilium
+// Copyright 2016-2020 Authors of Cilium
```
Comment (joestringer) [inline pkg/policy/api/kafka/doc.go]: nit: 
```suggestion
// Copyright 2017-2020 Authors of Cilium
```

**Review by joestringer — APPROVED** [review]
policy/docs bits LGTM. Awesome that we can reduce code AND add features here (add egress support). Probably even worth release-noting that.

**Comment by jrajahalme** [comment]
Raised a do-not-merge label, will add native protobufs for Kafka access logging and likely also for network policy as well to address the comments regarding hubble backwards compatibility and string constants.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

File: proxylib/kafka/parser.go (hunk 10 above)
Comment (jrajahalme) [reply proxylib/kafka/parser.go:105]: Did not, added. For now the errors use the same code as (policy) drops, will need to consider if that may cause confusion.

File: proxylib/kafka/parser.go (hunk 11 above)
Comment (jrajahalme) [reply proxylib/kafka/parser.go:130]: Yes, but I may have changed this since you reviewed.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/hubble/parser/seven/parser.go (hunk 9 above)
Comment (jrajahalme) [reply pkg/hubble/parser/seven/parser.go]: The most recent version got rid of the changes in hubble code, so this is not an issue any more :-)

**Comment by jrajahalme** [comment]
@nebril wrote:

> I think it's fine to go through with that, but would be nice to have a release note.

Added a release note.

>  If it's necessary to leave this, can we have information about Kafka in two places (AFAIU right now it will be stored in L7 field?).

Removed the generic L7 field use for Kafka, so info in one place again :-)

**Comment by jrajahalme** [comment]
test-me-please

**Comment by aanm** [comment]
CI failures seem to be related with the PR changes
```
0.00s)
PASS
coverage: 2.0% of statements in ./...
ok  	github.com/cilium/cilium/plugins/cilium-cni/types	0.011s	coverage: 2.0% of statements in ./...
?   	github.com/cilium/cilium/plugins/cilium-docker	[no test files]
?   	github.com/cilium/cilium/plugins/cilium-docker/driver	[no test files]
time="2020-07-28T12:13:57Z" level=info msg="init(): Registering kafkaParserFactory"
time="2020-07-28T12:13:57Z" level=info msg="proxylib: Initializing library"
=== RUN   TestMemcache
=== RUN   TestMemcache/text_set_pass
time="2020-07-28T12:13:57Z" level=fatal msg="Policy UnmarshalText failed: line 12: unknown field name \"l7_rules\" in cilium.L7NetworkPolicyRules"
FAIL	github.com/cilium/cilium/proxylib	0.021s
FAIL
Makefile:210: recipe for target 'unit-tests' failed
make: *** [unit-tests] Error 1
The command "./.travis/build.sh" exited with 2.
```

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
netnext hit an old (rate?) flake: #6307

**Comment by jrajahalme** [comment]
GKE failures also not related to Kafka.

**Comment by jrajahalme** [comment]
@aanm I removed the use of the generic L7 policy and access log formats and changed to use Kafka-specific formats instead. This removes the need for the string constants you referred to earlier and also reverted all changes in Hubble code.
There is also cleaner separation of the Cilium policy API Kafka structures (which are unchanged) and the the implementation type (now pkg/kafka.Rule) used for policy enforcement in the proxy code.
